Use CAIOSurface to cache IOSurfaces in the UI process.

mattwoodrow opened the pull request:
#### e95212f8d5a276cc28e4d36aa672172a11c066cf
<pre>
Use CAIOSurface to cache IOSurfaces in the UI process.
<a href="https://bugs.webkit.org/show_bug.cgi?id=253620">https://bugs.webkit.org/show_bug.cgi?id=253620</a>

Reviewed by Simon Fraser.

We can cache CAIOSurface objects for our layer buffers in the UI process.
This keeps CA rendering data alive, without marking the IOSurface as being in-use.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::decode):
(WebKit::RemoteLayerBackingStore::updateCachedBuffers):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:

Canonical link: <a href="https://commits.webkit.org/261553@main">https://commits.webkit.org/261553@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ad069f40334fde95e6492e1ea073568a1e66f0f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112108 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21240 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/732 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3905 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120765 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116170 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22588 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/12361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/4152 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117869 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16780 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99951 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105166 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98735 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/496 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45787 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13652 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/530 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/93408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14331 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/12361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19700 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52529 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8052 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16121 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/cocoa/IOSurface.h:
```diff
@@ -189,6 +191,8 @@ class IOSurface final {
 
     RetainPtr<IOSurfaceRef> m_surface;
 
+    mutable RetainPtr<id> m_caSurface;
```
I'd like a comment to say what the `id` really is (a `CAIOSurfaceRef`).

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
```diff
@@ -204,6 +208,11 @@ class RemoteLayerBackingStore {
     std::optional<MachSendRight> m_contentsBufferHandle;
+    RetainPtr<id> m_contentsBuffer;
+
+    std::optional<WebCore::RenderingResourceIdentifier> m_frontBufferIdentifier;
```
My review comment was going to be "Don't we have these already in `Buffer` classes above" but then I realize these are only used in the UI process. I think that's enough to justify either splitting `RemoteLayerBackingStore` into WP and UIP classes, or at least grouping the data members into structs which make it plain which process they are used it.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
```diff
@@ -87,6 +87,16 @@ class RemoteLayerTreeNode : public CanMakeWeakPtr<RemoteLayerTreeNode> {
 
+    struct CachedContentsBuffer {
+        WebCore::RenderingResourceIdentifier m_renderingResourceIdentifier;
+        RetainPtr<id> m_buffer;
```
Please add comment to say what the `id` really is.

smfr reviewed (approved):
Please look at the layout test crash.

mattwoodrow committed:
Use CAIOSurface to cache IOSurfaces in the UI process.
https://bugs.webkit.org/show_bug.cgi?id=253620

Reviewed by Simon Fraser.

We can cache CAIOSurface objects for our layer buffers in the UI process.
This keeps CA rendering data alive, without marking the IOSurface as being in-use.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::decode):
(WebKit::RemoteLayerBackingStore::updateCachedBuffers):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:

Canonical link: https://commits.webkit.org/261553@main
